[SPARK-26757][GraphX] Return 0 for `count` on empty Edge/Vertex RDDs

huonw opened the pull request:
## What changes were proposed in this pull request?

Previously a "java.lang.UnsupportedOperationException: empty
collection" exception would be thrown due to using `reduce`, rather
than `fold` or similar that can tolerate empty RDDs.

This behaviour has existed for the Vertex RDDs since it was introduced
in b30e0ae0351be1cbc0b1cf179293587b466ee026. It seems this behaviour
was inherited by the Edge RDDs via copy-paste in
ee29ef3800438501e0ff207feb00a28973fc0769.

## How was this patch tested?

Two new unit tests.

huonw committed:
[SPARK-26757][GraphX] Return 0 for `count` on empty Edge/Vertex RDDs

Previously a "java.lang.UnsupportedOperationException: empty
collection" exception would be thrown due to using `reduce`, rather
than `fold` or similar that can tolerate empty RDDs.

This behaviour has existed for the Vertex RDDs since it was introduced
in b30e0ae0351be1cbc0b1cf179293587b466ee026. It seems this behaviour
was inherited by the Edge RDDs via copy-paste in
ee29ef3800438501e0ff207feb00a28973fc0769.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #4537 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4537/testReport)** for PR 23681 at commit [`a843a0b`](https://github.com/apache/spark/commit/a843a0b8d8583ec885d354fef83a3f517a83c2a9).

SparkQA commented:
**[Test build #4537 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4537/testReport)** for PR 23681 at commit [`a843a0b`](https://github.com/apache/spark/commit/a843a0b8d8583ec885d354fef83a3f517a83c2a9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

maropu commented on graphx/src/main/scala/org/apache/spark/graphx/impl/EdgeRDDImpl.scala:
```diff
@@ -87,7 +87,7 @@ class EdgeRDDImpl[ED: ClassTag, VD: ClassTag] private[graphx] (
   /** The number of edges in the RDD. */
   override def count(): Long = {
-    partitionsRDD.map(_._2.size.toLong).reduce(_ + _)
+    partitionsRDD.map(_._2.size.toLong).fold(0)(_ + _)
```
The code below has the same problem in case of an empty edge?
https://github.com/apache/spark/blob/ae5b2a6a92be4986ef5b8062d7fb59318cff6430/graphx/src/main/scala/org/apache/spark/graphx/lib/SVDPlusPlus.scala#L75

srowen replied:
Oh nice catch, yeah. I searched for other "reduce" calls but didn't hit this one.

maropu commented on graphx/src/test/scala/org/apache/spark/graphx/EdgeRDDSuite.scala:
```diff
@@ -60,4 +60,15 @@ class EdgeRDDSuite extends SparkFunSuite with LocalSparkContext {
+      val empty = EdgeRDD.fromEdges(sc.emptyRDD[Edge[Int]])
+      assert(empty.count === 0)
+
+      val n = 100
```
`n` is not used.

huonw replied:
Oops, copy-paste. Thanks.

huonw committed:
[SPARK-26757][GraphX] Remove unused variable

huonw committed:
[SPARK-26757][GraphX] Make SVD++ work with no edges

Similar to `VertexRDD.count` and `EdgeRDD.count`, this used `reduce`
which fails on an empty RDD.

SparkQA commented:
**[Test build #4539 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4539/testReport)** for PR 23681 at commit [`de47631`](https://github.com/apache/spark/commit/de47631a0db84d285e4daafb5fdfffcace96a7c2).

SparkQA commented:
**[Test build #4539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4539/testReport)** for PR 23681 at commit [`de47631`](https://github.com/apache/spark/commit/de47631a0db84d285e4daafb5fdfffcace96a7c2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.
